File: solutions/scala/strain/1/src/main/scala/Strain.scala
object Strain {

  def keep[T](list: Seq[T], function: (T) => Boolean):Seq[T] = {
    list match {
      case f::rest => if (function(f)) Seq(f) ++ keep(rest, function) else keep(rest, function)
      case Nil => Nil
    }
  }

  def discard[T](list: Seq[T], function: (T) => Boolean) = {
    keep(list, (x:T) => !function(x))
  }

}

File: solutions/elixir/strain/1/lib/strain.ex
# For my Shree DR.MDD

defmodule Strain do
  @spec keep(list :: list(any), predicate :: (any -> boolean)) :: list(any)
  def keep(collection, predicate) do
    keep(collection, predicate, [])
  end

  defp keep([], _predicate, acc), do: Enum.reverse(acc)

  defp keep([head | tail], predicate, acc) do
    if predicate.(head) do
      keep(tail, predicate, [head | acc])
    else
      keep(tail, predicate, acc)
    end
  end

  @spec discard(list :: list(any), predicate :: (any -> boolean)) :: list(any)
  def discard(collection, predicate) do
    collection -- keep(collection, predicate)
  end
end
